document workflow to incrementally create a Kedro project (#4305)
kedro-org/kedro: docs/source/get_started/minimal_kedro_project.md

# Create a Minimal Kedro Project

> This documentation aims to explain the essential components of a minimal Kedro project. The guide begins with a blank project and gradually introduces the necessary elements. While most users typically start with a [project template]((./new_project.md)) or adapt an existing Python project, this guide will help you understand the core concepts and how to customise them to suit your specific needs.

Author: https://github.com/kedro-org/kedro/pull/4305#discussion_r1836815371

I think the idea is that, starting with a blank project isn't something that most people would do, so the goal of this documentation focus on explains the concept, but not a step-by-step guide that one should follow in reality.

Maybe I should flip the sentence like this?

While most users typically start with a [project template](./new_project.md)) or adapt an existing Python project, this guide begins with a blank project and gradually introduces the necessary elements. This will help you understand the core concepts and how to customise them to suit your specific needs.

| Directory/File        | Description                                                                 |
|-----------------------|-----------------------------------------------------------------------------|
| `conf/`               | Contains configuration files such as `catalog.yml` and `parameters.yml`.  |
| `data/`               | Local project data, typically not committed to version control.            |
| `docs/`               | Project documentation files.                                               |
| `notebooks/`          | Jupyter notebooks for experimentation and prototyping.                     |
| `src/`                | Source code for the project, including pipelines and nodes.                |
| `README.md`           | Project overview and instructions.                                         |
| `pyproject.toml`      | Metadata about the project, including dependencies.                        |
| `.gitignore`          | Specifies files and directories to be ignored by Git.                     |

### 2. **Mandatory Files**
For a project to be recognised as a Kedro project and support running `kedro run`, it must contain three essential files:

> - **`pyprojec.toml`**: Defines the python project

Reviewer: ```suggestion
- **`pyproject.toml`**: Defines the python project
```

- **`settings.py`**: Defines project settings, including library component registration.
- **`pipeline_registry.py`**: Registers the project's pipelines.

If you want to see some examples of these files, you can either create a project with `kedro new` or check out the [project template on GitHub](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights-pandas)

#### `pyproject.toml`

> The `pyproject.toml` file is a crucial component of a Kedro project, serving as the standard way to store build metadata and tool settings for Python projects. It is essential for defining the project's configuration and ensuring proper integration with various tools and libraries.

Reviewer: ```suggestion
The `pyproject.toml` file is a crucial component of a Kedro project that serves as the standard way to store build metadata and tool settings for Python projects. It is essential for defining the project's configuration and ensuring proper integration with various tools and libraries.
```

- Autodiscovery of Pipelines: Since Kedro 0.18.3, you can use the [`find_pipeline`](../nodes_and_pipelines/pipeline_registry.md#pipeline-autodiscovery) function to automatically discover pipelines defined in your project without manually updating the registry each time you create a new pipeline.

## Creating a Minimal Kedro Project Step-by-Step
This guide will walk you through the process of creating a minimal Kedro project, allowing you to successfully run `kedro run` with just three files.

### Step 1: Install Kedro

First, ensure that Python is installed on your machine. Then, install Kedro using pip:

```bash
pip install kedro
```

### Step 2: Create a New Kedro Project
Create a new directory for your project:
```bash
mkdir minikedro
```

Navigate into your newly created project directory:

```bash
cd minikiedro
```

### Step 3: Create `pyproject.toml`
Create a new file named `pyproject.toml` in the project directory with the following content:

```toml
[tool.kedro]
package_name = "minikedro"
project_name = "minikedro"
kedro_init_version = "0.19.9"
source_dir = "."

> ```

Reviewer 2: Interesting, so the Python packaging metadata (`[project]` table) is not even needed?

At this point, your workingn directory should look like this:
```bash
.
├── pyproject.toml
```

```{note}
Note we define `source_dir = "."`, usually we keep our source code inside a directory called `src`. For this example, we try to keep the structure minimal so we keep the source code in the root directory
```

### Step 4: Create `settings.py` and `pipeline_registry.py`
Next, create a folder named minikedro, which should match the package_name defined in pyproject.toml:

```bash
mkdir minikedro
```
Inside this folder, create two empty files: `settings.py` and `pipeline_registry.py`:

```bash
touch minikedro/settings.py minikedro/pipeline_registry.py
```

Now your working directory should look like this:
```bash
.
├── minikedro
│   ├── pipeline_registry.py
│   └── settings.py
└── pyproject.toml
```

Try running the following command in the terminal:
```bash
kedro run
```

> You will encounter an error indicating that pipeline_registry.py is empty:

Reviewer 3: ```suggestion
You will encounter an error indicating that `pipeline_registry.py` is empty:
```

```bash
AttributeError: module 'minikedro.pipeline_registry' has no attribute 'register_pipelines'
```

### Step 5: Create a Simple Pipeline
To resolve this issue, add the following code to `pipeline_registry.py`, which defines a simple pipeline to run:

```python
from kedro.pipeline import pipeline, node

def foo():
    return "dummy"

def register_pipelines():
    return {"__default__": pipeline([node(foo, None, "dummy_output")])}
```

If you attempt to run the pipeline again with `kedro run`, you will see another error:
```bash
MissingConfigException: Given configuration path either does not exist or is not a valid directory: /workspace/kedro/minikedro/conf/base
```

### Step 6: Define the Project Settings
This error occurs because Kedro expects a configuration folder named `conf`, along with two environments called `base` and `local`.

To fix this, add these two lines into `settings.py`:
```python
CONF_SOURCE = "."
CONFIG_LOADER_ARGS = {"base_env": ".", "default_run_env": "."}
```

> These lines override the default settings so that Kedro knows to look for configurations in the current directory instead of the expected conf folder. For more details, refer to [How to change the setting for a configuration source folder](../configuration/configuration_basics.md#how-to-change-the-setting-for-a-configuration-source-folder) and [Advance Configuration without a full Kedro project](../configuration/advanced_configuration.md#advanced-configuration-without-a-full-kedro-project)

Reviewer: ```suggestion
These lines override the default settings so that Kedro knows to look for configurations in the current directory instead of the expected `conf` folder. For more details, refer to [How to change the setting for a configuration source folder](../configuration/configuration_basics.md#how-to-change-the-setting-for-a-configuration-source-folder) and [Advance Configuration without a full Kedro project](../configuration/advanced_configuration.md#advanced-configuration-without-a-full-kedro-project)
```
